### PR TITLE
New serverless pattern - apigw-websocket-api-lambda-authorizer

### PR DESCRIPTION
**Pull request by ankushjain358**
*Issue #, if available:*

*Description of changes:* New serverless pattern **WebSocket API Cognito Authentication using Lambda Authorizer**

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: apigw-websocket-api-lambda-authorizer/Readme.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
+# WebSocket API Cognito Authentication using Lambda Authorizer
+
+This pattern demonstrates how to integrate Cognito authentication with Amazon API Gateway WebSocket API.
```
Comment (bfreiberg) [inline apigw-websocket-api-lambda-authorizer/Readme.md:3]:
```suggestion
This pattern demonstrates how to integrate Amazon Cognito authentication with Amazon API Gateway WebSocket API.
```

File: apigw-websocket-api-lambda-authorizer/Readme.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    * Enter a stack name
+    * Enter the desired AWS Region
+    * Allow SAM CLI to create IAM roles with the required permissions.
+
+    Once you have run `sam deploy -guided` mode and saved arguments to a configuration file (samconfig.toml), you can use `sam deploy` in future to use these defaults.
+
+5. Note the outputs from the SAM deployment process. These contain the WebSocketURI, UserpoolId, and ClientId which are used for testing.
+
+## Testing
+In order to test, follow the steps given below:
```
Comment (bfreiberg) [inline apigw-websocket-api-lambda-authorizer/Readme.md:48]: Please include a note about the importance of strong passwords for production workloads and that this only serves as a simple demo

File: apigw-websocket-api-lambda-authorizer/template.yaml [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+      CodeUri: './src/WebSocketAPI'
+      MemorySize: 256
+      Timeout: 30
+      Policies:
+        - DynamoDBCrudPolicy:
+            TableName: !Ref 'ConnectionMappingTableName'
+        - Statement:
+            Effect: Allow
+            Action: execute-api:ManageConnections
+            Resource: arn:aws:execute-api:*:*:*/@connections/*
```
Comment (bfreiberg) [inline apigw-websocket-api-lambda-authorizer/template.yaml:80]: Close scope down this policy a bit, use pseudo parameters for the accountId and region: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/pseudo-parameter-reference.html

File: apigw-websocket-api-lambda-authorizer/src/WebSocketAPI.sln
Comment (bfreiberg) [inline apigw-websocket-api-lambda-authorizer/src/WebSocketAPI.sln:1]: Can this file be deleted?

File: apigw-websocket-api-lambda-authorizer/src/WebSocketAPI/Readme.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
+# WebSocket API Project
+
+This starter project consists of:
+* serverless.template - an AWS CloudFormation Serverless Application Model template file for declaring your Serverless functions and other AWS resources
```
Comment (bfreiberg) [inline apigw-websocket-api-lambda-authorizer/src/WebSocketAPI/Readme.md:4]: this line seems superfluous

File: apigw-websocket-api-lambda-authorizer/src/LambdaAuthorizer/Readme.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
+# AWS Lambda Empty Function Project
+
+This starter project consists of:
```
Comment (bfreiberg) [inline apigw-websocket-api-lambda-authorizer/src/LambdaAuthorizer/Readme.md:3]: Please add a line regarding CognitoJwtVerifier.cs as well

File: apigw-websocket-api-lambda-authorizer/src/WebSocketAPI.sln
Comment (ankushjain358) [reply apigw-websocket-api-lambda-authorizer/src/WebSocketAPI.sln:1]: This is an important file for .NET based project. The solution maintains the state information for projects. We should not delete it. Refer [Solution (.sln) file](https://learn.microsoft.com/en-us/visualstudio/extensibility/internals/solution-dot-sln-file?view=vs-2022)

File: apigw-websocket-api-lambda-authorizer/src/WebSocketAPI/Readme.md (hunk 4 above)
Comment (ankushjain358) [reply apigw-websocket-api-lambda-authorizer/src/WebSocketAPI/Readme.md:4]: The project level README files are auto-generated when we create a project using Visual Studio. So, I have not touched it. Instead, I have created a main README file at root level.

File: apigw-websocket-api-lambda-authorizer/Readme.md (hunk 1 above)
Comment (ankushjain358) [reply apigw-websocket-api-lambda-authorizer/Readme.md]: Fixed

File: apigw-websocket-api-lambda-authorizer/Readme.md (hunk 2 above)
Comment (ankushjain358) [reply apigw-websocket-api-lambda-authorizer/Readme.md:48]: Done

File: apigw-websocket-api-lambda-authorizer/src/LambdaAuthorizer/Readme.md (hunk 5 above)
Comment (ankushjain358) [reply apigw-websocket-api-lambda-authorizer/src/LambdaAuthorizer/Readme.md:3]: Done

File: apigw-websocket-api-lambda-authorizer/template.yaml (hunk 3 above)
Comment (ankushjain358) [reply apigw-websocket-api-lambda-authorizer/template.yaml]: Done

**Comment by ankushjain358** [comment]
@bfreiberg - I have made the requested changes. Can you please verify?

**Comment by bfreiberg** [comment]
Looks good, thank you for your contribution. The PR will be merged soon.

File: apigw-websocket-api-lambda-authorizer/apigw-websocket-api-lambda-authorizer.json [hunk 6]
```diff
@@ -0,0 +1,100 @@
+{
+    "title": "WebSocket API Cognito Authentication using Lambda Authorizer",
```
Comment (bfreiberg) [inline apigw-websocket-api-lambda-authorizer/apigw-websocket-api-lambda-authorizer.json:2]: Please change to "Amazon Cognito authentication with Amazon API Gateway Websockets"

File: apigw-websocket-api-lambda-authorizer/apigw-websocket-api-lambda-authorizer.json [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    "description": "This pattern demonstrates how to integrate Cognito authentication with Amazon API Gateway WebSocket API.",
```
Comment (bfreiberg) [inline apigw-websocket-api-lambda-authorizer/apigw-websocket-api-lambda-authorizer.json:3]: Please change to "How to integrate Cognito authentication with Amazon API Gateway WebSocket API."

File: apigw-websocket-api-lambda-authorizer/Readme.md [hunk 8]
```diff
@@ -0,0 +1,118 @@
+# WebSocket API Cognito Authentication using Lambda Authorizer
+
+This pattern demonstrates how to integrate Amazon Cognito authentication with Amazon API Gateway WebSocket API.
+
+It includes the Lambda implementations for Lambda Authorizer, Lambda functions for $connect, $disconnect and custom route, and AWS Serverless Application Model (SAM) code to deploy backend infrastructure.
+
+Learn more about this pattern at Serverless Land Patterns: https://serverlessland.com/patterns/apigw-websocket-api-lambda-authorizer
+
+## Why Lambda Authorizer, not Cognito Authorizer?
+API Gateway Websocket API doesn't support [Cognito authorizer](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-enable-cognito-user-pool.html) as of now. To enable Cognito Authentication for API Gateway Websocket API you need to create a Lambda authorizer instead of using the built-in Cognito authorizer.
```
Comment (bfreiberg) [inline apigw-websocket-api-lambda-authorizer/Readme.md:10]: Please change "Authentication" to "authentication"

File: apigw-websocket-api-lambda-authorizer/apigw-websocket-api-lambda-authorizer.json [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    "language": ".NET",
+    "level": "200",
+    "framework": "SAM",
+    "introBox": {
+        "headline": "How it works",
+        "text": [
+            "This pattern demonstrates how to integrate Cognito authentication with Amazon API Gateway WebSocket API.",
+            "It includes the Lambda implementations for Lambda Authorizer, Lambda functions for $connect, $disconnect and custom route, and AWS Serverless Application Model (SAM) code to deploy backend infrastructure."
```
Comment (bfreiberg) [inline apigw-websocket-api-lambda-authorizer/apigw-websocket-api-lambda-authorizer.json:11]: Please change "Authorizer" to "authorizer"

File: apigw-websocket-api-lambda-authorizer/apigw-websocket-api-lambda-authorizer.json [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    },
+    "deploy": {
+        "text": [
+            "sam build",
+            "sam deploy --guided"
+        ]
+    },
+    "testing": {
+        "text": [
+            "See the Github repo for detailed testing instructions."
```
Comment (bfreiberg) [inline apigw-websocket-api-lambda-authorizer/apigw-websocket-api-lambda-authorizer.json:31]: please change "Github" to "GitHub"

File: apigw-websocket-api-lambda-authorizer/apigw-websocket-api-lambda-authorizer.json [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        ]
+    },
+    "cleanup": {
+        "text": [
+            "Delete the stack: sam delete"
```
Comment (bfreiberg) [inline apigw-websocket-api-lambda-authorizer/apigw-websocket-api-lambda-authorizer.json:36]: please change to "<code>sam delete</code>"

**Review by bfreiberg — COMMENTED** [review]
I'm sorry but there a couple of spelling issues that need to be adjusted.

**Comment by bfreiberg** [comment]
Hi @ankushjain358, you likely see it already on #2048 but the changes look good and should be merged soon

**Pull request merged**